File: src/com/maddyhome/idea/vim/helper/SearchHighlightsHelper.java
/*
 * IdeaVim - Vim emulator for IDEs based on the IntelliJ platform
 * Copyright (C) 2003-2020 The IdeaVim authors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package com.maddyhome.idea.vim.helper;

import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.EditorFactory;
import com.intellij.openapi.editor.colors.EditorColors;
import com.intellij.openapi.editor.colors.EditorColorsScheme;
import com.intellij.openapi.editor.markup.*;
import com.intellij.openapi.fileEditor.FileEditorManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.project.ProjectManager;
import com.intellij.ui.ColorUtil;
import com.maddyhome.idea.vim.common.TextRange;
import com.maddyhome.idea.vim.ex.ranges.LineRange;
import com.maddyhome.idea.vim.group.SearchGroup;
import com.maddyhome.idea.vim.option.OptionsManager;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.awt.*;
import java.util.*;
import java.util.List;

public class SearchHighlightsHelper {
  public static void updateSearchHighlights(@Nullable String pattern, boolean shouldIgnoreSmartCase, boolean showHighlights, boolean forceUpdate) {
    updateSearchHighlights(pattern, shouldIgnoreSmartCase, showHighlights, -1, null, true, forceUpdate);
  }

  public static int updateIncsearchHighlights(@NotNull Editor editor, @NotNull String pattern, boolean forwards, int caretOffset, @Nullable LineRange searchRange) {
    final int searchStartOffset = searchRange != null ? EditorHelper.getLineStartOffset(editor, searchRange.startLine) : caretOffset;
    final boolean showHighlights = OptionsManager.INSTANCE.getHlsearch().isSet();
    return updateSearchHighlights(pattern, false, showHighlights, searchStartOffset, searchRange, forwards, false);
  }

  @NotNull
  public static RangeHighlighter addSubstitutionConfirmationHighlight(@NotNull Editor editor, int start, int end) {
    TextAttributes color = new TextAttributes(
      editor.getColorsScheme().getColor(EditorColors.SELECTION_FOREGROUND_COLOR),
      editor.getColorsScheme().getColor(EditorColors.SELECTION_BACKGROUND_COLOR),
      editor.getColorsScheme().getColor(EditorColors.CARET_COLOR),
      EffectType.ROUNDED_BOX, Font.PLAIN
    );
    return editor.getMarkupModel().addRangeHighlighter(start, end, HighlighterLayer.SELECTION,
      color, HighlighterTargetArea.EXACT_RANGE);
  }

  /**
   * Refreshes current search highlights for all editors of currently active text editor/document
   */
  private static int updateSearchHighlights(@Nullable String pattern, boolean shouldIgnoreSmartCase, boolean showHighlights,
                                            int initialOffset, @Nullable LineRange searchRange, boolean forwards, boolean forceUpdate) {
    int currentMatchOffset = -1;

    ProjectManager projectManager = ProjectManager.getInstanceIfCreated();
    if (projectManager == null) return currentMatchOffset;
    Project[] projects = projectManager.getOpenProjects();
    for (Project project : projects) {
      Editor current = FileEditorManager.getInstance(project).getSelectedTextEditor();
      Editor[] editors = current == null ? null : EditorFactory.getInstance().getEditors(current.getDocument(), project);
      if (editors == null) {
        continue;
      }

      for (final Editor editor : editors) {
        // Try to keep existing highlights if possible. Update if hlsearch has changed or if the pattern has changed.
        // Force update for the situations where the text is the same, but the ignore case values have changed.
        // E.g. Use `*` to search for a word (which ignores smartcase), then use `/<Up>` to search for the same pattern,
        // which will match smartcase. Or changing the smartcase/ignorecase settings
        if (shouldRemoveSearchHighlights(editor, pattern, showHighlights) || forceUpdate) {
          removeSearchHighlights(editor);
        }

        if (shouldAddAllSearchHighlights(editor, pattern, showHighlights)) {
          // hlsearch (+ incsearch/noincsearch)
          final int startLine = searchRange == null ? 0 : searchRange.startLine;
          final int endLine = searchRange == null ? -1 : searchRange.endLine;
          List<TextRange> results = SearchGroup.findAll(editor, pattern, startLine, endLine,
            SearchHelper.shouldIgnoreCase(pattern, shouldIgnoreSmartCase));
          if (!results.isEmpty()) {
            currentMatchOffset = findClosestMatch(editor, results, initialOffset, forwards);
            highlightSearchResults(editor, pattern, results, currentMatchOffset);
          }
          UserDataManager.setVimLastSearch(editor, pattern);
        }
        else if (shouldAddCurrentMatchSearchHighlight(pattern, showHighlights, initialOffset)) {
          // nohlsearch + incsearch
          final boolean wrap = OptionsManager.INSTANCE.getWrapscan().isSet();
          final EnumSet<SearchGroup.SearchOptions> searchOptions = EnumSet.of(SearchGroup.SearchOptions.WHOLE_FILE);
          if (wrap) searchOptions.add(SearchGroup.SearchOptions.WRAP);
          if (shouldIgnoreSmartCase) searchOptions.add(SearchGroup.SearchOptions.IGNORE_SMARTCASE);
          if (!forwards) searchOptions.add(SearchGroup.SearchOptions.BACKWARDS);
          final TextRange result = SearchGroup.findIt(editor, pattern, initialOffset, 1, searchOptions);
          if (result != null && pattern != null) {
            currentMatchOffset = result.getStartOffset();
            final List<TextRange> results = Collections.singletonList(result);
            highlightSearchResults(editor, pattern, results, currentMatchOffset);
          }
        }
        else if (shouldMaintainCurrentMatchOffset(pattern, initialOffset)) {
          // incsearch. If nothing has changed (e.g. we've edited offset values in `/foo/e+2`) make sure we return the
          // current match offset so the caret remains at the current incsarch match
          final Integer offset = UserDataManager.getVimIncsearchCurrentMatchOffset(editor);
          if (offset != null) {
            currentMatchOffset = offset;
          }
        }
      }
    }

    return currentMatchOffset;
  }

  /**
   * Remove current search highlights if hlSearch is false, or if the pattern is changed
   */
  @Contract("_, _, false -> true; _, null, true -> false")
  private static boolean shouldRemoveSearchHighlights(@NotNull Editor editor, @Nullable String newPattern, boolean hlSearch) {
    return !hlSearch || (newPattern != null && !newPattern.equals(UserDataManager.getVimLastSearch(editor)));
  }

  private static void removeSearchHighlights(@NotNull Editor editor) {
    UserDataManager.setVimLastSearch(editor, null);

    Collection<RangeHighlighter> ehl = UserDataManager.getVimLastHighlighters(editor);
    if (ehl == null) {
      return;
    }

    for (RangeHighlighter rh : ehl) {
      editor.getMarkupModel().removeHighlighter(rh);
    }

    ehl.clear();

    UserDataManager.setVimLastHighlighters(editor, null);
  }

  /**
   * Add search highlights if hlSearch is true and the pattern is changed
   */
  @Contract("_, _, false -> false; _, null, true -> false")
  private static boolean shouldAddAllSearchHighlights(@NotNull Editor editor, @Nullable String newPattern, boolean hlSearch) {
    return hlSearch && newPattern != null && !newPattern.equals(UserDataManager.getVimLastSearch(editor)) && !Objects.equals(newPattern, "");
  }

  private static int findClosestMatch(@NotNull Editor editor, @NotNull List<TextRange> results, int initialOffset, boolean forwards) {
    if (results.isEmpty() || initialOffset == -1) {
      return -1;
    }

    final int size = EditorHelperRt.getFileSize(editor);
    final TextRange max = Collections.max(results, (r1, r2) -> {
      final int d1 = distance(r1, initialOffset, forwards, size);
      final int d2 = distance(r2, initialOffset, forwards, size);
      if (d1 < 0 && d2 >= 0) {
        return Integer.MAX_VALUE;
      }
      return d2 - d1;
    });

    if (!OptionsManager.INSTANCE.getWrapscan().isSet()) {
      final int start = max.getStartOffset();
      if (forwards && start < initialOffset) {
        return -1;
      }
      else if (start >= initialOffset) {
        return -1;
      }
    }

    return max.getStartOffset();
  }

  private static int distance(@NotNull TextRange range, int pos, boolean forwards, int size) {
    final int start = range.getStartOffset();
    if (start <= pos) {
      return forwards ? size - pos + start : pos - start;
    }
    else {
      return forwards ? start - pos : pos + size - start;
    }
  }

  public static void highlightSearchResults(@NotNull Editor editor, @NotNull String pattern, List<TextRange> results,
                                            int currentMatchOffset) {
    Collection<RangeHighlighter> highlighters = UserDataManager.getVimLastHighlighters(editor);
    if (highlighters == null) {
      highlighters = new ArrayList<>();
      UserDataManager.setVimLastHighlighters(editor, highlighters);
    }

    for (TextRange range : results) {
      final boolean current = range.getStartOffset() == currentMatchOffset;
      final RangeHighlighter highlighter = highlightMatch(editor, range.getStartOffset(), range.getEndOffset(), current, pattern);
      highlighters.add(highlighter);
    }

    UserDataManager.setVimIncsearchCurrentMatchOffset(editor, currentMatchOffset);
  }

  private static @NotNull RangeHighlighter highlightMatch(@NotNull Editor editor, int start, int end, boolean current, String tooltip) {
    TextAttributes attributes = editor.getColorsScheme().getAttributes(EditorColors.TEXT_SEARCH_RESULT_ATTRIBUTES);
    if (current) {
      // This mimics what IntelliJ does with the Find live preview
      attributes = attributes.clone();
      attributes.setEffectType(EffectType.ROUNDED_BOX);
      attributes.setEffectColor(editor.getColorsScheme().getColor(EditorColors.CARET_COLOR));
    }
    if (attributes.getErrorStripeColor() == null) {
      attributes.setErrorStripeColor(getFallbackErrorStripeColor(attributes, editor.getColorsScheme()));
    }
    final RangeHighlighter highlighter = editor.getMarkupModel().addRangeHighlighter(start, end, HighlighterLayer.SELECTION - 1,
      attributes, HighlighterTargetArea.EXACT_RANGE);
    highlighter.setErrorStripeTooltip(tooltip);
    return highlighter;
  }

  /**
   * Return a valid error stripe colour based on editor background
   *
   * <p>Based on HighlightManager#addRangeHighlight behaviour, which we can't use because it will hide highlights
   * when hitting Escape.</p>
   */
  private static @Nullable Color getFallbackErrorStripeColor(TextAttributes attributes, EditorColorsScheme colorsScheme) {
    if (attributes.getBackgroundColor() != null) {
      boolean isDark = ColorUtil.isDark(colorsScheme.getDefaultBackground());
      return isDark ? attributes.getBackgroundColor().brighter() : attributes.getBackgroundColor().darker();
    }
    return null;
  }

  /**
   * Add search highlight for current match if hlsearch is false and we're performing incsearch highlights
   */
  @Contract("_, true, _ -> false")
  private static boolean shouldAddCurrentMatchSearchHighlight(@Nullable String pattern, boolean hlSearch, int initialOffset) {
    return !hlSearch && isIncrementalSearchHighlights(initialOffset) && pattern != null && pattern.length() > 0;
  }

  /**
   * Keep the current match offset if the pattern is still valid and we're performing incremental search highlights
   * This will keep the caret position when editing the offset in e.g. `/foo/e+1`
   */
  @Contract("null, _ -> false")
  private static boolean shouldMaintainCurrentMatchOffset(@Nullable String pattern, int initialOffset) {
    return pattern != null && pattern.length() > 0 && isIncrementalSearchHighlights(initialOffset);
  }

  /**
   * initialOffset is only valid if we're highlighting incsearch
   */
  @Contract(pure = true)
  private static boolean isIncrementalSearchHighlights(int initialOffset) {
    return initialOffset != -1;
  }
}

File: src/main/java/com/maddyhome/idea/vim/helper/SearchHighlightsHelper.kt
/*
 * Copyright 2003-2026 The IdeaVim authors
 *
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE.txt file or at
 * https://opensource.org/licenses/MIT.
 */

@file:JvmName("SearchHighlightsHelper")

package com.maddyhome.idea.vim.helper

import com.intellij.openapi.editor.Editor
import com.intellij.openapi.editor.colors.EditorColors
import com.intellij.openapi.editor.markup.EffectType
import com.intellij.openapi.editor.markup.HighlighterLayer
import com.intellij.openapi.editor.markup.HighlighterTargetArea
import com.intellij.openapi.editor.markup.RangeHighlighter
import com.intellij.openapi.editor.markup.TextAttributes
import com.maddyhome.idea.vim.api.VimEditor
import com.maddyhome.idea.vim.api.globalOptions
import com.maddyhome.idea.vim.api.injector
import com.maddyhome.idea.vim.api.options
import com.maddyhome.idea.vim.common.TextRange
import com.maddyhome.idea.vim.ex.ranges.LineRange
import com.maddyhome.idea.vim.newapi.IjVimEditor
import com.maddyhome.idea.vim.newapi.ij
import com.maddyhome.idea.vim.newapi.vim
import com.maddyhome.idea.vim.state.mode.inCommandLineModeWithVisual
import com.maddyhome.idea.vim.state.mode.inVisualMode
import org.jetbrains.annotations.Contract
import java.awt.Font
import java.util.*

fun updateSearchHighlights(
  pattern: String?,
  shouldIgnoreSmartCase: Boolean,
  showHighlights: Boolean,
  forceUpdate: Boolean,
) {
  updateSearchHighlights(null, pattern, 1, shouldIgnoreSmartCase, showHighlights, -1, null, true, forceUpdate)
}

fun updateSearchCount(
  pattern: String?,
  shouldIgnoreSmartCase: Boolean,
  currentMatchOffset: Int = -1,
) {
  if (pattern == null) return
  val selectedEditor = injector.editorGroup.getSelectedEditor() ?: return
  updateSearchCount(
    selectedEditor,
    pattern,
    0,
    selectedEditor.lineCount() - 1,
    shouldIgnoreSmartCase,
    currentMatchOffset,
    selectedEditor.ij,
  )
}

fun updateIncsearchHighlights(
  editor: Editor,
  pattern: String,
  count1: Int,
  forwards: Boolean,
  caretOffset: Int,
  searchRange: LineRange?,
  forceShowAllMatches: Boolean = false,
): Int {
  val searchStartOffset = if (searchRange != null && searchRange.startLine < editor.document.lineCount) {
    editor.vim.getLineStartOffset(searchRange.startLine)
  } else {
    caretOffset
  }
  val showHighlights = injector.options(editor.vim).hlsearch || forceShowAllMatches
  return updateSearchHighlights(
    editor.vim,
    pattern,
    count1,
    false,
    showHighlights,
    searchStartOffset,
    searchRange,
    forwards,
    false
  )
}

fun addSubstitutionConfirmationHighlight(editor: Editor, start: Int, end: Int): RangeHighlighter {
  val color = TextAttributes(
    editor.colorsScheme.getColor(EditorColors.SELECTION_FOREGROUND_COLOR),
    editor.colorsScheme.getColor(EditorColors.SELECTION_BACKGROUND_COLOR),
    editor.colorsScheme.getColor(EditorColors.CARET_COLOR),
    EffectType.ROUNDED_BOX,
    Font.PLAIN,
  )
  return editor.markupModel.addRangeHighlighter(
    start,
    end,
    HighlighterLayer.SELECTION,
    color,
    HighlighterTargetArea.EXACT_RANGE,
  )
}

/**
 * Highlight a single range using the standard search-result attributes, returning the highlighter so the caller can
 * remove it later.
 *
 * Unlike the search highlights, this is not tracked by [SearchHighlights], so it is suitable for transient overlays -
 * such as the `inccommand` preview - that manage their own highlighter lifecycle.
 */
fun highlightPreviewMatch(editor: Editor, start: Int, end: Int, tooltip: String): RangeHighlighter {
  return createSearchMatchHighlighter(editor, start, end, tooltip)
}

/**
 * Refreshes current search highlights for all visible editors
 */
private fun updateSearchHighlights(
  currentEditor: VimEditor?,
  pattern: String?,
  count1: Int,
  shouldIgnoreSmartCase: Boolean,
  showHighlights: Boolean,
  initialOffset: Int,
  searchRange: LineRange?,
  forwards: Boolean,
  forceUpdate: Boolean,
): Int {
  var currentEditorCurrentMatchOffset = -1

  // Update highlights in all visible editors. We update non-visible editors when they get focus.
  // Note that this now includes all editors - main, diff windows, even toolwindows like the Commit editor and consoles
  val editors = injector.editorGroup.getEditors().filter {
    !it.ij.isDisposed
      && (injector.application.isUnitTest() || it.ij.component.isShowing)
      && (currentEditor == null || it.projectId == currentEditor.projectId)
  }

  val isIncsearch = isIncrementalSearchHighlights(initialOffset)

  editors.forEach { vimEditor ->
    val editor = vimEditor.ij
    val isCurrentEditor = editor === currentEditor?.ij

    // Try to keep existing highlights if possible. Update if hlsearch has changed or if the pattern has changed.
    // Force update for the situations where the text is the same, but the ignore case values have changed.
    // E.g., Use `*` to search for a word (which ignores smartcase), then use `/<Up>` to search for the same pattern,
    // which will match smartcase. Or changing the smartcase/ignorecase settings
    if (shouldRemoveSearchHighlights(editor, pattern, showHighlights) || forceUpdate) {
      removeSearchHighlights(editor)
    }

    if (pattern == null) return@forEach

    val searchStartLine = searchRange?.startLine ?: 0
    val searchEndLine = (searchRange?.endLine ?: -1).coerceAtMost(vimEditor.lineCount() - 1)

    // The current match belongs to the 'incsearch' preview, and only to the editor being searched. Once the search has
    // been accepted or cancelled there is no current match any more, however the caret is then moved - the result is
    // simply where the caret is, and every match looks the same again (VIM-4308)
    val currentMatch = if (isIncsearch && isCurrentEditor) {
      findIncsearchMatch(vimEditor, pattern, initialOffset, count1, forwards, shouldIgnoreSmartCase)
    } else {
      null
    }
    val currentMatchOffset = currentMatch?.startOffset ?: -1

    if (shouldAddAllSearchHighlights(pattern, showHighlights)) {
      // hlsearch (+ incsearch/noincsearch)
      addAllSearchHighlights(vimEditor, pattern, searchStartLine, searchEndLine, shouldIgnoreSmartCase)
      editor.vimLastSearch = pattern
    } else if (currentMatch != null && shouldAddCurrentMatchSearchHighlight(pattern, showHighlights, initialOffset)) {
      // nohlsearch + incsearch. Even though search highlights are disabled, we still show a highlight (current editor
      // only), because 'incsearch' is active. We don't show one if Visual is active (behind Command-line, of course),
      // because the Visual selection is enough
      if (!vimEditor.inVisualMode && !vimEditor.inCommandLineModeWithVisual) {
        editor.vimSearchHighlights.showSingleMatch(pattern, currentMatch)
      }
    }

    editor.vimSearchHighlights.setCurrentMatch(currentMatchOffset)

    // Remember the incsearch match, so that `c_CTRL-R_CTRL-W` can insert the word after it
    editor.vimIncsearchCurrentMatch = currentMatch

    if (isCurrentEditor) {
      currentEditorCurrentMatchOffset = currentMatchOffset
    }

    updateSearchCount(
      vimEditor,
      pattern,
      searchStartLine,
      searchEndLine,
      shouldIgnoreSmartCase,
      currentMatchOffset,
      editor
    )
  }

  return currentEditorCurrentMatchOffset
}

private fun addAllSearchHighlights(
  editor: VimEditor,
  pattern: String,
  searchStartLine: Int,
  searchEndLine: Int,
  shouldIgnoreSmartCase: Boolean,
) {
  // Make sure the range fits this editor. Note that Vim will use the same range for all windows. E.g., given
  // `:1,5s/foo`, Vim will highlight all occurrences of `foo` in the first five lines of all visible windows
  if (searchStartLine > editor.lineCount() - 1) return

  editor.ij.vimSearchHighlights.ensureUpToDate(
    pattern,
    shouldIgnoreCase(pattern, shouldIgnoreSmartCase),
    searchStartLine,
    searchEndLine,
  )
}

/**
 * Find the match that the in-progress search would move to - Vim's `hl-CurSearch`, and what `c_CTRL-G`/`c_CTRL-T`
 * step through. Returns null if the pattern doesn't match.
 *
 * This is a real search rather than a lookup among the highlighted matches. The highlights only cover part of the
 * document, and are only rebuilt when something about them changes, but the current match moves while the pattern
 * stays the same.
 */
private fun findIncsearchMatch(
  editor: VimEditor,
  pattern: String,
  initialOffset: Int,
  count1: Int,
  forwards: Boolean,
  shouldIgnoreSmartCase: Boolean,
): TextRange? {
  val searchOptions = EnumSet.of(SearchOptions.WHOLE_FILE)
  if (injector.globalOptions().wrapscan) searchOptions.add(SearchOptions.WRAP)
  if (shouldIgnoreSmartCase) searchOptions.add(SearchOptions.IGNORE_SMARTCASE)
  if (!forwards) searchOptions.add(SearchOptions.BACKWARDS)

  return injector.searchHelper.findPattern(editor, pattern, initialOffset, count1, searchOptions)
}

fun updateSearchCount(
  vimEditor: VimEditor,
  pattern: String,
  searchStartLine: Int,
  searchEndLine: Int,
  shouldIgnoreSmartCase: Boolean,
  currentMatchOffset: Int,
  editor: Editor,
) {
  val matchOffset = if (currentMatchOffset != -1) currentMatchOffset else editor.caretModel.offset
  val maxSearchCount = injector.globalOptions().maxsearchcount
  val searchCount = computeSearchCount(
    vimEditor,
    pattern,
    searchStartLine,
    searchEndLine,
    shouldIgnoreCase(pattern, shouldIgnoreSmartCase),
    matchOffset,
    maxSearchCount,
  )
  if (searchCount.current <= 0) {
    return
  }
  injector.outputPanel.getOrCreate(
    IjVimEditor(editor),
    injector.executionContextManager.getEditorExecutionContext(IjVimEditor(editor))
  ).statusText = formatSearchCountText(searchCount, maxSearchCount)
}

private data class SearchCount(val current: Int, val total: Int, val exceededMaxCount: Boolean)

private fun computeSearchCount(
  vimEditor: VimEditor,
  pattern: String,
  searchStartLine: Int,
  searchEndLine: Int,
  ignoreCase: Boolean,
  matchOffset: Int,
  maxSearchCount: Int,
): SearchCount {
  val maxMatchesToFind = if (maxSearchCount > 0) maxSearchCount + 1 else Int.MAX_VALUE
  val results = injector.searchHelper.findAll(
    vimEditor,
    pattern,
    searchStartLine,
    searchEndLine,
    ignoreCase,
    maxMatchesToFind,
  )
  var current = 0
  var total = 0
  var exceededMaxCount = false
  for (range in results) {
    total++
    if (range.startOffset <= matchOffset) {
      current = total
    }
    if (maxSearchCount > 0 && total > maxSearchCount) {
      exceededMaxCount = true
      break
    }
  }
  if (current == 0 && total > 0) {
    current = total
  }
  return SearchCount(current, total, exceededMaxCount)
}

private fun formatSearchCountText(searchCount: SearchCount, maxSearchCount: Int): String {
  val current = searchCount.current
  val total = searchCount.total
  if (!searchCount.exceededMaxCount || total <= maxSearchCount) {
    return "[$current/$total]"
  }
  return if (current > maxSearchCount) {
    "[>$maxSearchCount/>$maxSearchCount]"
  } else {
    "[$current/>$maxSearchCount]"
  }
}

/**
 * Remove current search highlights if hlSearch is false, or if the pattern is changed
 */
@Contract("_, _, false -> true; _, null, true -> false")
private fun shouldRemoveSearchHighlights(editor: Editor, newPattern: String?, hlSearch: Boolean): Boolean {
  return !hlSearch || newPattern != null && newPattern != editor.vimLastSearch
}

private fun removeSearchHighlights(editor: Editor) {
  editor.vimLastSearch = null
  editor.vimSearchHighlights.clear()
}

/**
 * Add search highlights if hlSearch is true and there is a pattern to highlight.
 *
 * There is deliberately no check for the pattern being unchanged - [SearchHighlights.ensureUpToDate] compares the
 * request against what the editor is already showing, so asking for highlights that are already there costs nothing.
 */
@Contract("_, false -> false; null, true -> false")
private fun shouldAddAllSearchHighlights(newPattern: String?, hlSearch: Boolean): Boolean {
  return hlSearch && !newPattern.isNullOrEmpty()
}

/**
 * Remove the current match highlight, e.g. after the caret has moved or the document has been edited. Does nothing if
 * there are no search highlights.
 *
 * The highlight belongs to the 'incsearch' preview and to nothing else, so there is never anything to work out from the
 * caret - moving it just takes the highlight away. While the command line is open, though, the preview owns the
 * highlight and follows the search and `c_CTRL-G`/`c_CTRL-T` rather than the caret, so reapply it instead.
 */
fun clearCurrentSearchMatchHighlight(editor: Editor) {
  if (editor.isDisposed) return
  editor.vimSearchHighlights.setCurrentMatch(editor.vimIncsearchCurrentMatch?.startOffset ?: -1)
}

/**
 * Add search highlight for current match if hlsearch is false, and we're performing incsearch highlights
 */
@Contract("_, true, _ -> false")
private fun shouldAddCurrentMatchSearchHighlight(pattern: String?, hlSearch: Boolean, initialOffset: Int): Boolean {
  return !hlSearch && isIncrementalSearchHighlights(initialOffset) && !pattern.isNullOrEmpty()
}

/**
 * initialOffset is only valid if we're highlighting incsearch
 */
@Contract(pure = true)
private fun isIncrementalSearchHighlights(initialOffset: Int) = initialOffset != -1
